src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -1338,28 +1350,72 @@ public static void readAdditionalLevelSaveData(LevelStorageSource.LevelDirectory
+            final var mismatchEvent = new ModMismatchEvent(levelDirectory, mismatchedVersions, missingVersions);
             ModLoader.get().postEvent(mismatchEvent);
 
+            StringBuilder resolved = new StringBuilder("The following mods have version differences that were marked resolved:");
+            StringBuilder unresolved = new StringBuilder("The following mods have version differences that were not resolved:");
+
+            // For mods that were marked resolved, log the version resolution and the mod that resolved the mismatch
+            mismatchEvent.getResolved().forEachOrdered((res) ->
+            {
+                final var modid = res.modid();
+                final var diff = res.versionDifference();
+                if(res.wasSelfResolved()) {
+                    resolved.append(System.lineSeparator())
+                            .append(diff.isMissing() ?
+                                    "%s (version %s -> MISSING, self-resolved)".formatted(modid, diff.oldVersion()) :
+                                    "%s (version %s -> %s, self-resolved)".formatted(modid, diff.oldVersion(), diff.newVersion())
+                            );
+                }
+                else
+                {
+                    final var resolver = res.resolver().getModId();
+                    resolved.append(System.lineSeparator())
+                            .append(diff.isMissing() ?
+                                    "%s (version %s -> MISSING, resolved by %s)".formatted(modid, diff.oldVersion(), resolver) :
+                                    "%s (version %s -> %s, resolved by %s)".formatted(modid, diff.oldVersion(), diff.newVersion(), resolver)
+                            );
+                }
+            });
+
             // For mods that did not specify handling, show a warning to users that errors may occur
-            mismatchedVersions.keySet().stream()
-                    .filter(modId -> mismatchEvent.getModState(modId) == ModVersionsMismatchEvent.MismatchHandlingState.UNHANDLED)
-                    .forEach(modId -> {
-                        LOGGER.warn(WORLDPERSISTENCE, "This world was saved with mod {} version {} and it is now at version {}, things may not work well",
-                                modId, mismatchedVersions.get(modId), mismatchEvent.getCurrentVersion(modId));
-                    });
+            mismatchEvent.getUnresolved().forEachOrdered((unres) ->
+            {
+                final var modid = unres.modid();
+                final var diff = unres.versionDifference();
+                unresolved.append(System.lineSeparator())
+                        .append(diff.isMissing() ?
+                                "%s (version %s -> MISSING)".formatted(modid, diff.oldVersion()) :
+                                "%s (version %s -> %s)".formatted(modid, diff.oldVersion(), diff.newVersion())
+                        );
+            });
+
+            if(mismatchEvent.anyResolved())
```

Comment:
Formatting: space between `if` keyword and parenthesis.
```suggestion
            if (mismatchEvent.anyResolved())
```